### PR TITLE
Introduce required/mandatory filters

### PR DESCRIPTION
**Pull request by nishantmonu51**
There is no way to mark a filter as **mandatory**: a canvas always renders, even if a critical filter (e.g. `region`, `customer_id`) has been left empty. Mandatory/Required ensures that a user always filters on one or more dimension values. 

This PR adds support for declaring required filters on canvas dashboards, alongside the existing `pinned` filters.

  - New `filters.required` YAML field on canvas dashboards
  - Required filters are implicitly pinned and cannot be removed from the filter row.
  - If any required filter has no value (from `defaults.filters` or the user),  the canvas body is replaced with an inline panel listing the missing filters; the filter row stays interactive so the user can set values.
  - Filter pills for unsatisfied required filters render in a red error state with a tooltip explaining the requirement.
  - Enforcement applies in view, embed, and editor modes — authors see the same blocked state their viewers will see.
  - `runtime/ai/instructions/data/resources/canvas.md` updated so the analyst and developer agents know about the new field.
 
Screenshot 2026-05-23 at 2 52 02 AM

**Checklist:**
- [ ] Covered by tests
- [ ] Ran it and it works as intended
- [ ] Reviewed the diff before requesting a review
- [ ] Checked for unhandled edge cases
- [ ] Linked the issues it closes
- [ ] Checked if the docs need to be updated. If so, create a separate Linear DOCS issue
- [ ] Intend to cherry-pick into the release branch
- [ ] I'm proud of this work!

### CODE REVIEW

File: runtime/parser/parse_canvas.go [hunk 1]
```diff
@@ -288,6 +289,7 @@ func (p *Parser) parseCanvas(node *Node) error {
 	r.CanvasSpec.Rows = rows
 	r.CanvasSpec.SecurityRules = rules
 	r.CanvasSpec.PinnedFilters = tmp.Filters.Pinned
+	r.CanvasSpec.RequiredFilters = tmp.Filters.Required
```
Comment (AdityaHegde) [inline runtime/parser/parse_canvas.go:292]: We should have validation for this to ensure it is defined in metrics view's measure/dimensions. Since there is no no-code way to set it a validation will help with issues.

File: web-common/src/features/dashboards/filters/dimension-filters/DimensionFilter.svelte [hunk 2]
```diff
@@ -452,9 +455,17 @@
...
+                >{required
+                  ? "required dimension"
+                  : "dimension"}</svelte:fragment
+              >
             </TooltipTitle>
-            Click to edit the the filters in this dimension
+            {#if missingRequired}
+              This filter is required. Select a value to load the dashboard.
+            {:else}
+              Click to edit the the filters in this dimension
```
Comment (AdityaHegde) [inline web-common/src/features/dashboards/filters/dimension-filters/DimensionFilter.svelte]: nit: remove doubled `the`

File: web-common/src/features/canvas/stores/canvas-entity.ts [hunk 3]
```diff
@@ -359,10 +362,13 @@ export class CanvasEntity {
...
 
     const pinnedFilters = get(this.filterManager.pinnedFilterKeysStore);
+    const requiredFilters = get(this.filterManager.requiredFilterKeysStore);
 
-    const genericPinnedKeys = Array.from(pinnedFilters).map(
-      (f) => f.split("::")[1],
-    );
+    // Required filters are implicitly pinned at runtime; persist them only under
+    // `filters.required` to avoid duplicating them in `filters.pinned`.
+    const genericPinnedKeys = Array.from(pinnedFilters)
```
Comment (AdityaHegde) [inline web-common/src/features/canvas/stores/canvas-entity.ts]: We should do this dedupe render time instead. When we add a toggle for required switching it on and off will removed the pinned filter.

**Review by AdityaHegde — CHANGES_REQUESTED** [review]
Is a UI toggle for required filter a future thing? We have one for pin in the filter dropdown.

File: runtime/parser/parse_canvas.go (hunk 1 above)
Comment (nishantmonu51) [reply runtime/parser/parse_canvas.go:292]: Added validation at reconcile time in `canvas.go` (`validateRequiredFilters`) — checks each `filters.required` entry exists as a dimension or measure on at least one metrics view referenced by the canvas' components. Did it at the reconciler rather than the parser since dimensions live on a separate resource the parser can't see; the reconciler already has the metrics view specs in hand.

File: web-common/src/features/dashboards/filters/dimension-filters/DimensionFilter.svelte (hunk 2 above)
Comment (nishantmonu51) [reply web-common/src/features/dashboards/filters/dimension-filters/DimensionFilter.svelte]: Fixed.

File: web-common/src/features/canvas/stores/canvas-entity.ts (hunk 3 above)
Comment (nishantmonu51) [reply web-common/src/features/canvas/stores/canvas-entity.ts]: Moved. Pinned and required are now tracked independently in memory; `convertToUIFilters` treats a filter as visible when either is set, and the sort orders required + pinned ahead of everything else. `saveDefaultFilters` writes both lists as-is, so toggling required off no longer drops the pin.

**Comment by nishantmonu51** [comment]
@AdityaHegde re: the UI toggle question — added a `*` toggle next to the pin button in the dimension and measure filter dropdowns. Builder-only like pin, and round-trips to `filters.required` on Save as default.

**Review by AdityaHegde — APPROVED** [review]
